src/core/model/types/bitset.h

```diff
@@ -0,0 +1,1190 @@
+        pt = static_cast<CharT*>(alloca(NumBits));
+    } else {
+        pt = new CharT[NumBits];
+    }
+    Buffer const buf(pt);
+
+    CharType const zero = is.widen('0');
+    CharType const one = is.widen('1');
+
+    typename IOSBase::iostate state = IOSBase::goodbit;
+    typename IStreamType::sentry sentry(is);
+    if (sentry) {
+        try {
+            for (size_t i{NumBits}; i > 0; --i) {
+                static typename Traits::int_type eof = Traits::eof();
+
+                typename Traits::int_type c1 = is.rdbuf()->sbumpc();
+                if (Traits::eq_int_type(c1, eof)) {
+                    state |= IOSBase::eofbit;
+                    break;
+                } else {
+                    CharType const c2 = Traits::to_char_type(c1);
+                    if (Traits::eq(c2, zero)) {
+                        *pt++ = zero;
+                    } else if (Traits::eq(c2, one)) {
+                        *pt++ = one;
+                    } else if (Traits::eq_int_type(is.rdbuf()->sputbackc(c2), eof)) {
+                        state |= IOSBase::failbit;
+                        break;
+                    }
+                }
+            }
+        } catch (...) {
+            is.setstate(IOSBase::badbit);
+        }
+    }
+
+    if constexpr (NumBits) {
+        if (size_t len = pt - buf.ptr) {
+            x.template CopyFromPtr<CharT, Traits>(buf.ptr, len, 0, len, zero, one);
+        } else {
+            state |= IOSBase::failbit;
+        }
+    }
+    if (state) {
+        is.setstate(state);
+    }
+    return is;
+}
+
+template <typename CharT, typename Traits, size_t NumBits>
+std::basic_ostream<CharT, Traits>& operator<<(std::basic_ostream<CharT, Traits>& os,
+                                              BitsetImpl<NumBits> const& x) {
+    std::basic_string<CharT, Traits> tmp;
+
+    std::ctype<CharT> const& ct = std::use_facet<std::ctype<CharT>>(os.getloc());
+    x.CopyToString(tmp, ct.widen('0'), ct.widen('1'));
+    return os << tmp;
+}
+
+template <typename Bitset>
+concept HasExtesnsions = requires(Bitset bs) {
+    bs._Find_first();
+    bs._Find_next(0);
+};
+
+template <size_t S>
+struct BitsetSelector {
+    using Type = BitsetImpl<S>;
+};
+
+template <size_t S>
+    requires HasExtesnsions<std::bitset<S>>
+struct BitsetSelector<S> {
+    using Type = std::bitset<S>;
+};
+
+}  // namespace bitset_impl
+
+template <size_t S>
+using Bitset = typename bitset_impl::BitsetSelector<S>::Type;
+
+}  // namespace model
+
+namespace std {
+
+// For hash to be named properly
+// NOLINTBEGIN(readability-identifier-naming)
+/// @note This hash is equal to hash of @b libc++ std::bitset with the same data.
+/// It's @b not equal to hash of @b libstdc++ std::bitset with the same data.
+template <size_t NumBits>
+struct hash<model::bitset_impl::BitsetImpl<NumBits>> {
+    size_t operator()(model::bitset_impl::BitsetImpl<NumBits> const& b) const noexcept {
+        return b.HashCode();
+    }
+};
```

Comment:
It is undefined behavior to add declarations or definitions to namespace std or to any namespace nested within std.

We can add template specializations, but this should be done without opening namespace std with `namespace` keyword:

```cpp
template<>
struct std::hash<MyType>
{
    std::size_t operator()(const MyType& t) const { return t.hash(); }
};
```